alibi/prototypes/protoselect.py

```diff
@@ -0,0 +1,543 @@
+        data['prototypes_labels'] = np.concatenate([[l] * len(protos[l])
+                                                    for l in protos]).astype(np.int32)  # noqa: E741
+        data['prototypes'] = self.Y[data['prototypes_indices']]
+        return Explanation(meta=self.meta, data=data)
+
+    def save(self, path: Union[str, os.PathLike]) -> None:
+        super().save(path)
+
+    @classmethod
+    def load(cls, path: Union[str, os.PathLike]) -> "Explainer":  # type: ignore[override]
+        return super().load(path, predictor=None)
+
+
+def _helper_protoselect_euclidean_1knn(explainer: ProtoSelect,
+                                       num_prototypes: int,
+                                       eps: float,
+                                       preprocess_fn: Optional[Callable[[np.ndarray], np.ndarray]] = None
+                                       ) -> Optional[KNeighborsClassifier]:
+    """
+    Helper function to fit a 1-KNN classifier on the prototypes returned by the explainer.
+    Sets the epsilon radius to be used.
+
+    Parameters
+    ----------
+    explainer
+        Fitted explainer.
+    num_prototypes
+        Number of requested prototypes.
+    eps
+        Epsilon radius to be set and used for the computation of prototypes.
+     preprocess_fn
+        Preprocessing function to be applied to the data instance before fitting the 1-KNN classifier
+
+    Returns
+    -------
+    Fitted KNN-classifier.
+    """
+    # update explainer eps and get explanation
+    explainer.eps = eps
+    explanation = explainer.explain(num_prototypes=num_prototypes)
+
+    # train 1-knn classifier
+    proto, proto_labels = explanation.data['prototypes'], explanation.data['prototypes_labels']
+    if len(proto) == 0:
+        return None
+
+    knn = KNeighborsClassifier(n_neighbors=1)
+    return knn.fit(X=proto, y=proto_labels)
+
+
+def cv_protoselect_euclidean(refset: Tuple[np.ndarray, np.ndarray],
+                             protoset: Tuple[np.ndarray, ],
+                             valset: Optional[Tuple[np.ndarray, np.ndarray]] = None,
+                             num_prototypes: int = 1,
+                             eps_range: Optional[np.ndarray] = None,
+                             quantiles: Optional[Tuple[float, float]] = None,
+                             grid_size: int = 25,
+                             n_splits: int = 2,
+                             batch_size: int = int(1e10),
+                             preprocess_fn: Optional[Callable[[np.ndarray], np.ndarray]] = None,
+                             **kwargs) -> dict:
+    """
+    Cross-validation parameter selection for ProtoSelect with Euclidean distance. The method computes
+    the best epsilon radius.
+
+    Parameters
+    ----------
+    refset
+        Tuple, `(X_ref, X_ref_labels)`, consisting of the reference data instances with the corresponding reference
```

Comment:
To discuss - consistent naming between this and `ProtoSelect.fit` where it's `X` and `X_labels`.

Reply:
I've renamed them to be consistent with the suggestions for the fit function.